### PR TITLE
feat(posfilter): Create loaders for wave glider L1 data for roll-pitch-heading and their std-devs

### PR DESCRIPTION
**Pull request by madhavmk**
* New loader "read_raw_l1_data_files()" for reading L1 data file efficiently using regex pattern matching INSPVAA / INSSTDDEVA data formats.
* gps_ws_time_to_j2000_time() function to convert GPS weeks and seconds to seconds from J2000 epoch
* Add unittests to improve coverage 
* Closes Issue #226

### CODE REVIEW

File: src/gnatss/loaders.py [hunk 1]
```diff
@@ -440,3 +447,56 @@ def load_quality_control(qc_files: List[str], time_scale="tt") -> pd.DataFrame:
         qc_df = pd.DataFrame(columns=[constants.QC_STARTTIME, constants.QC_ENDTIME])
 
     return qc_df
+
+
+def read_raw_l1_data_files(
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:452]:
```suggestion
def read_novatel_L1_data_files(
```
I think we should be a little more specific for clarity

File: src/gnatss/loaders.py [hunk 2]
```diff
@@ -440,3 +447,56 @@ def load_quality_control(qc_files: List[str], time_scale="tt") -> pd.DataFrame:
...
+    data_files: list[str], data_format: Literal["INSPVAA", "INSSTDEVA"] = "INSPVAA"
+) -> pd.DataFrame:
+    """
+    Read from L1 data files and return its dataframe representation
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:456]: Please add links to the Novatel website
INSSTDEVA: https://docs.novatel.com/OEM7/Content/SPAN_Logs/INSSTDEV.htm?tocpath=Commands%20%2526%20Logs%7CLogs%7CSPAN%20Logs%7C_____30
INSPVAA: https://docs.novatel.com/OEM7/Content/SPAN_Logs/INSPVA.htm?tocpath=Commands%20%2526%20Logs%7CLogs%7CSPAN%20Logs%7C_____22

File: src/gnatss/loaders.py [hunk 3]
```diff
@@ -440,3 +447,56 @@ def load_quality_control(qc_files: List[str], time_scale="tt") -> pd.DataFrame:
...
+        data_file_arrays.append(data_file_array)
+    all_data_array = np.concatenate(data_file_arrays, axis=0, casting="no")
+
+    df = pd.DataFrame(all_data_array)
+    df[constants.RPH_TIME] = pd.Series(
+        gps_ws_time_to_j2000_time(all_data_array["Week"], all_data_array["Seconds"])
+    )
+
+    if data_format == "INSPVAA":
+        df = df.loc[df["Status"] == "INS_SOLUTION_GOOD"]
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:497]: This is great! 🙌🏽

File: src/gnatss/constants/__init__.py [hunk 4]
```diff
@@ -108,3 +108,75 @@
...
+            "float",
+            "float",
+            "float",
+            "float",
+            "float",
+            "object",
+        ),
+    },
+    "INSSTDEVA": {
+        "regex_pattern": r"#INSSTDEVA,.*?,.*?,.*?,.*?,(.*?),(.*?),.*?,.*?,.*?;(.*?),(.*?),"
```
Comment (lsetiawan) [inline src/gnatss/constants/__init__.py:148]: This is so clever! Looking good on parsing week and seconds from header with regex 👍🏽

File: src/gnatss/loaders.py [hunk 5]
```diff
@@ -440,3 +447,56 @@ def load_quality_control(qc_files: List[str], time_scale="tt") -> pd.DataFrame:
...
+) -> pd.DataFrame:
+    """
+    Read from L1 data files and return its dataframe representation
+
+    Parameters
+    ----------
+    data_files: list[str]
+        Paths to the L1 data files to be loaded
+
+    data_format: Literal["INSPVAA", "INSSTDEVA"]
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:463]:
```suggestion
    data_format: {"INSPVAA", "INSSTDEVA"}
```

File: src/gnatss/loaders.py [hunk 6]
```diff
@@ -440,3 +447,56 @@ def load_quality_control(qc_files: List[str], time_scale="tt") -> pd.DataFrame:
...
+
+def read_raw_l1_data_files(
+    data_files: list[str], data_format: Literal["INSPVAA", "INSSTDEVA"] = "INSPVAA"
+) -> pd.DataFrame:
+    """
+    Read from L1 data files and return its dataframe representation
+
+    Parameters
+    ----------
+    data_files: list[str]
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:460]:
```suggestion
    data_files: list of str
```

File: src/gnatss/loaders.py [hunk 7]
```diff
@@ -440,3 +447,56 @@ def load_quality_control(qc_files: List[str], time_scale="tt") -> pd.DataFrame:
...
+        Paths to the L1 data files to be loaded
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:461]:
```suggestion
        Paths to the Novatel L1 data files to be loaded
```

File: src/gnatss/loaders.py [hunk 8]
```diff
@@ -440,3 +447,56 @@ def load_quality_control(qc_files: List[str], time_scale="tt") -> pd.DataFrame:
...
+        data_file_text = Path(data_file).read_text()
+        all_groups = re_pattern.findall(data_file_text)
+        data_fields_dtypes = zip(
+            l1_data_config.get("data_fields"),
+            l1_data_config.get("data_fields_dtypes"),
+        )
+
+        data_file_array = np.array(all_groups, dtype=list(data_fields_dtypes))
+        data_file_arrays.append(data_file_array)
+    all_data_array = np.concatenate(data_file_arrays, axis=0, casting="no")
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:489]: Could you please summarize what's happening here with a comment?

File: src/gnatss/loaders.py [hunk 9]
```diff
@@ -440,3 +447,56 @@ def load_quality_control(qc_files: List[str], time_scale="tt") -> pd.DataFrame:
...
+
+    df = pd.DataFrame(all_data_array)
+    df[constants.RPH_TIME] = pd.Series(
```
Comment (lsetiawan) [inline src/gnatss/loaders.py:492]:
```suggestion
    # convert GNSS Week and Seconds to J2000 Seconds
    df[constants.RPH_TIME] = pd.Series(
```

File: src/gnatss/utilities/time.py [hunk 10]
```diff
@@ -26,38 +26,39 @@ class TimeUnixJ2000(TimeFromEpoch):
...
     # Get the week start time exactly at midnight, doesn't include leap seconds, TT scale
     week_start = AstroTime(week_time.strftime("%Y-%m-%d"), format="iso", scale="tt")
+
     # Add seconds to beginning of week
-    final_time = week_start + TimeDelta(seconds, format="sec")
-    return final_time
+    final_time = week_start[unique_index] + TimeDelta(seconds, format="sec")
+
+    final_time_j2000 = final_time.unix_j2000
+    return final_time_j2000
```
Comment (lsetiawan) [inline src/gnatss/utilities/time.py:64]: While this is great that you're abstracting getting the j2000 and renaming the function... but can you revert this back to be more general and return AstroTime, this way, it's not tied to just j2000. You can always do a lambda function to get the unix_j2000 or something.

File: src/gnatss/loaders.py (hunk 3 above)
Comment (madhavmk) [reply src/gnatss/loaders.py:498]: Thanks :smile:

File: src/gnatss/loaders.py (hunk 9 above)
Comment (madhavmk) [reply src/gnatss/loaders.py:493]: This is being done in line 498

**Comment by madhavmk** [comment]
Hello @lsetiawan ,
Thank you for the valuable PR comments. I have addressed them.

**Pull request merged**